core/mount/mount_idmapped_utils_linux_go122.go

```diff
@@ -0,0 +1,88 @@
+//go:build !go1.23 && linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/containerd/containerd/v2/pkg/sys"
+
+	"golang.org/x/sys/unix"
+)
+
+// getUsernsFD returns pinnable user namespace's file descriptor.
+func getUsernsFD(uidMaps, gidMaps []syscall.SysProcIDMap) (*os.File, error) {
+	var pidfd int
+
+	proc, err := os.StartProcess("/proc/self/exe", []string{"containerd[getUsernsFD]"}, &os.ProcAttr{
+		Sys: &syscall.SysProcAttr{
+			Cloneflags:  unix.CLONE_NEWUSER,
+			UidMappings: uidMaps,
+			GidMappings: gidMaps,
+			// NOTE: It's reexec but it's not heavy because subprocess
+			// be in PTRACE_TRACEME mode before performing execve.
+			Ptrace:    true,
+			Pdeathsig: syscall.SIGKILL,
+			PidFD:     &pidfd,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to start noop process for unshare: %w", err)
+	}
+
+	if pidfd == -1 || !sys.SupportsPidFD() {
```

Comment:
nit: I'd check for pidfd support at the very beginning, and return an error if it doesn't. As you do in the go 1.23 file

Reply:
Let me check it in old kernel which doesn't support CLONE_PIDFD. If the kernel doesn't support CLONE_PIDFD, the `os.StartProcess` should return error. If so, we don't need to check pidfd value.

Reply:
Confirmed that the os.StartProcess won't return error if the kernel doesn't support CLONE_PIDFD.

```
vagrant@vagrant:~$ uname -a
Linux vagrant 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
vagrant@vagrant:~$ sudo /vagrant/test -test.v -test.root -test.run TestIdmappedMount/IDMapMount
=== RUN   TestIdmappedMount
=== RUN   TestIdmappedMount/IDMapMount
pidfd =  -1
    mount_idmapped_linux_test.go:130:
                Error Trace:    /home/fuweid/go/src/github.com/containerd/containerd/core/mount/mount_idmapped_linux_test.go:130
                Error:          Received unexpected error:
                                failed to prevent pid reused issue because pidfd isn't supported
                Test:           TestIdmappedMount/IDMapMount
--- FAIL: TestIdmappedMount (0.00s)
    --- FAIL: TestIdmappedMount/IDMapMount (0.00s)
FAIL
```

I applied the following patch based on this pull request.

```diff
diff --git a/core/mount/mount_idmapped_linux_test.go b/core/mount/mount_idmapped_linux_test.go
index 90e3a61e8..be802705a 100644
--- a/core/mount/mount_idmapped_linux_test.go
+++ b/core/mount/mount_idmapped_linux_test.go
@@ -24,7 +24,6 @@ import (
        "syscall"
        "testing"

-       kernel "github.com/containerd/containerd/v2/pkg/kernelversion"
        "github.com/containerd/continuity/testutil"
        "github.com/stretchr/testify/require"
 )
@@ -74,12 +73,14 @@ var (
 func TestIdmappedMount(t *testing.T) {
        testutil.RequiresRoot(t)

-       k512 := kernel.KernelVersion{Kernel: 5, Major: 12}
-       ok, err := kernel.GreaterEqualThan(k512)
-       require.NoError(t, err)
-       if !ok {
-               t.Skip("GetUsernsFD requires kernel >= 5.12")
-       }
+       /*
+               k512 := kernel.KernelVersion{Kernel: 5, Major: 12}
+               ok, err := kernel.GreaterEqualThan(k512)
+               require.NoError(t, err)
+               if !ok {
+                       t.Skip("GetUsernsFD requires kernel >= 5.12")
+               }
+       */

        t.Run("GetUsernsFD", testGetUsernsFD)

diff --git a/core/mount/mount_idmapped_utils_linux_go122.go b/core/mount/mount_idmapped_utils_linux_go122.go
index 21b1c81a3..a0b9546e4 100644
--- a/core/mount/mount_idmapped_utils_linux_go122.go
+++ b/core/mount/mount_idmapped_utils_linux_go122.go
@@ -48,7 +48,8 @@ func getUsernsFD(uidMaps, gidMaps []syscall.SysProcIDMap) (*os.File, error) {
                return nil, fmt.Errorf("failed to start noop process for unshare: %w", err)
        }

-       if pidfd == -1 || !sys.SupportsPidFD() {
+       if pidfd == -1 {
+               fmt.Println("pidfd = ", pidfd)
                proc.Kill()
                proc.Wait()
                return nil, fmt.Errorf("failed to prevent pid reused issue because pidfd isn't supported")

```

Reply:
> As you do in the go 1.23 file

 Since go1.23.0 has pidfd double-close issue, I check the `SupportsPidFD` at the beginning.
But for the go1.22.0, I need to check `pidfd` result. So I combine two condition checks together.

Reply:
Ohhh, I see now. Makse sense, thanks!